src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.Icu.cs

```diff
@@ -12,77 +12,6 @@ internal partial class CultureData
         // ICU constants
         private const int ICU_ULOC_KEYWORD_AND_VALUES_CAPACITY = 100; // max size of keyword or value
         private const int ICU_ULOC_FULLNAME_CAPACITY = 157;           // max size of locale name
-        private const string ICU_COLLATION_KEYWORD = "@collation=";
-
-        /// <summary>
-        /// This method uses the sRealName field (which is initialized by the constructor before this is called) to
-        /// initialize the rest of the state of CultureData based on the underlying OS globalization library.
-        /// </summary>
-        private unsafe bool IcuInitCultureData()
-        {
-            Debug.Assert(_sRealName != null);
-
-            Debug.Assert(!GlobalizationMode.Invariant);
-            Debug.Assert(!GlobalizationMode.UseNls);
-
-            string realNameBuffer = _sRealName;
-
-            // Basic validation
-            if (realNameBuffer.Contains('@'))
-            {
-                return false; // don't allow ICU variants to come in directly
-            }
-
-            // Replace _ (alternate sort) with @collation= for ICU
-            ReadOnlySpan<char> alternateSortName = default;
-            int index = realNameBuffer.IndexOf('_');
-            if (index > 0)
-            {
-                if (index >= (realNameBuffer.Length - 1) // must have characters after _
-                    || realNameBuffer.IndexOf('_', index + 1) >= 0) // only one _ allowed
-                {
-                    return false; // fail
-                }
-                alternateSortName = realNameBuffer.AsSpan(index + 1);
-                realNameBuffer = string.Concat(realNameBuffer.AsSpan(0, index), ICU_COLLATION_KEYWORD, alternateSortName);
```

Comment:
Unrelated to the change at hand, but does this just take the NLS alternate sort names and blindly tack them after the "`@collation`" keyword? 

If so, then FWIW, this won't work as the NLS alternate sort names do not match the ICU keywords.

For example:
In NLS you'll have "`de-DE_phoneb`", but with ICU this would need to be "`de-DE@collation=phonebook`" or "`de-DE-u-co-phonebk`"

Reply:
Same remark on the ICU->NLS side as well I think.
(You'll end up with `de-DE_phonebook` when you probably wanted `de-DE_phoneb`).

Reply:
Isn't the user is passing us the name? this is not conversion between Nls and ICU. This is just normalizing the locale name to make it work with ICU. We'll just respect whatever the user giving us. 

The question is, if anyone want to use a locale name that can work on Windows(NLS) and ICU, what name they should use? we can ty to help in converting the alternatesortname but I am seeing this can cause other problems as users will get something different than what they passed to us. what you think about that?

---
In reply to: [418672754](https://github.com/dotnet/runtime/pull/35680#discussion_r418672754) [](ancestors = 418672754)

Reply:
So it seems like in this case we will endup with `de-DE@collation=phoneb` for ICU:
https://github.com/dotnet/runtime/blob/9351765ba806a40b4da4bfae876c8f398a6ffc97/src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs#L870

That calls into: https://github.com/dotnet/runtime/blob/1db45ed953745c0e270d09df99828cb58772f4ad/src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.Icu.cs#L312

Which would return: 
https://github.com/dotnet/runtime/blob/1db45ed953745c0e270d09df99828cb58772f4ad/src/libraries/System.Private.CoreLib/src/System/Globalization/IcuLocaleData.cs#L4440-L4442

And that returns: https://github.com/dotnet/runtime/blob/1db45ed953745c0e270d09df99828cb58772f4ad/src/libraries/System.Private.CoreLib/src/System/Globalization/IcuLocaleData.cs#L154

NLS scenario PInvokes to get the local name  from the locale id so I get we're fine on NLS: https://github.com/dotnet/runtime/blob/f185bd5571f2500b4843d145418c4418747246b6/src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.Nls.cs#L610

cc: @tarekgh am I following this correct?

Reply:
@safern this is not about mapping LCID to culture name.

Reply:
>Isn't the user is passing us the name? this is not conversion between Nls and ICU. This is just normalizing the locale name to make it work with ICU

Well in that case why bother handling the `_` (alt sort) at all then? If this is just taking the user input then I guess it's up to them to give the correct input I suppose? 
If the intent isn't to convert between NLS and ICU at all, then perhaps maybe add a comment to that effect -- as the code is somewhat confusing (at least to me seeing this for the first time). :)

Reading a comment like this:
`// Replace _ (alternate sort) with @collation= for ICU`
and then seeing a bunch of code that alters/manipulates the input locale name gives the impression that you're trying to do at least *some* kind of mapping from the given input to an ICU style locale name.

> if anyone want to use a locale name that can work on Windows(NLS) and ICU, what name they should use?
I'd probably suggest using a BCP47 language tag like this: "de-DE-u-co-phonebk", though at least for Windows(NLS) that would only work on Windows 10, and not down-level, so its not a great solution.

> but I am seeing this can cause other problems as users will get something different than what they passed to us. 
Yes, that's a fair point. I don't know what the expectation is w.r.t. altering the given input tags at all.

Reply:
> Well in that case why bother handling the _ (alt sort) at all then? If this is just taking the user input then I guess it's up to them to give the correct input I suppose?

.NET doesn't support ICU naming convention. Users have to give the expected form of locale names. We used the original Windows convention for naming when using alternative sort (i.e. `_`). To call ICU, we'll have normalize the name to the form that ICU can work nicely with. that is why we do this normalization.

> I'd probably suggest using a BCP47 language tag like this: "de-DE-u-co-phonebk", though at least for Windows(NLS) that would only work on Windows 10, and not down-level, so its not a great solution.

This will not work on downlevel as you pointed out.

Reply:
Anyways, my intent was not to derail this PR with unrelated issues... :)

This was mostly just wanting to point something out that I noticed while looking at the change. I was somewhat surprised to see this kind of manipulation of locale tags, and I (mistakenly) thought the code was trying to do some kind of mapping here, and getting it wrong, so I wanted to try to help out by pointing it out. It might be worth discussing this offline later on I suppose.

Reply:
@jefgen your feedback is always valuable and adds something to our knowledge. keep doing that :-). We can discuss that offline as you suggested.

Reply:
Thanks @tarekgh. :-)

One last comment I suppose...
> We used the original Windows convention for naming when using alternative sort (i.e. _).

Hmm, well if this is the case (using Windows-style naming), then *maybe* you might want to consider some kind of basic mapping of the names? (In order to have `de-DE_phoneb` input still give phonebook sort order).
But then again, I'm not sure if this is perhaps needed at all, given this:
> Users have to give the expected form of locale names.

So maybe this is a broader question of expectations/documentation on what sorts of input tags should be used, etc.

Reply:
> then maybe you might want to consider some kind of basic mapping of the names?

We may consider that as long as we are not changing the name that the user passed to us. I mean CultureInfo.Name property will match the name that user passed to us. I think this will be reasonable to do. does the phonebook sorting name is the only case here or we have more cases like that?